lhotse/dataset/sampling.py

```diff
@@ -0,0 +1,288 @@
+import logging
+import math
+import random
+import warnings
+from abc import ABC, abstractmethod
+from typing import Iterable, List, Optional, Tuple
+
+import torch
+
+from lhotse import CutSet
+
+
+class CutSampler(ABC):
+    """
+    CutSampler is responsible for collecting batches of cuts, given specified criteria.
+    It implements correct handling of multiprocessing in DataLoader,
+    as well as distributed training, so that the cuts are not duplicated across workers.
+
+    Sampling in a CutSampler is intended to be very quick - it only uses the metadata in
+    ``CutSet`` manifest to select the cuts, and is not intended to perform any I/O.
+
+    .. note:
+
+        For implementers of new samplers:
+        Subclasses of CutSampler are expected to implement ``__next__()`` to introduce specific
+        sampling logic (e.g. based on filters such as max number of frames/tokens/etc.).
+        CutSampler defines ``__iter__()`` which optionally shuffles the cut IDs, and resets
+        ``self.current_idx`` to zero (to be used and incremented inside of ``__next__()``.
+    """
+
+    def __init__(
+            self,
+            cut_ids: Iterable[str],
+            shuffle: bool = False,
+            world_size: int = 1,
+            local_rank: int = 0,
+    ) -> None:
+        """
+
+        :param cut_ids: An iterable of cut IDs for the full dataset.
+            CutSampler will take care of partitioning that into parallel/distributed workers.
+        :param shuffle: When ``True``, the cuts will be shuffled at the start of iteration.
+            Convenient when mini-batch loop is inside an outer epoch-level loop, e.g.:
+            `for epoch in range(10): for batch in dataset: ...` as every epoch will see a
+            different cuts order.
+        :param world_size: Total number of distributed nodes. Set only when using ``DistributedDataParallel``.
+        :param local_rank: Index of distributed node. Set only when using ``DistributedDataParallel``.
+        """
+        self.all_cut_ids = list(cut_ids)
+        self.shuffle = shuffle
+        self.world_size = world_size
+        self.local_rank = local_rank
+
+    def __iter__(self) -> 'CutSampler':
+        """
+        Prepare the dataset for iterating over a new epoch. Will shuffle the data if requested.
+        """
+        # We can only retrieve the cut ids once iter is called, not before!
+        # Iter is called inside a dataloader process, whereas init is called when
+        # instantiating this class (typically before dataloader).
+        self.cut_ids = partition_cut_ids(self.all_cut_ids, world_size=self.world_size, local_rank=self.local_rank)
+        if self.shuffle:
+            random.shuffle(self.cut_ids)
+        self.current_idx = 0
+        return self
+
+    @abstractmethod
```

Comment:
This is a pythonic way of saying `virtual CutSet next() = 0` in C++; it will raise an exception saying that the method is abstract when you try to run it in the base class. That's why this class also inherits from `ABC`, you need both that (or setting the metaclass to `ABCMeta`) and the `abstractmethod` decorator to get this behavior.

Reply:
Oh, and the `...` in the place of the function's body is equivalent to writing `pass`; in some codebases, I noticed that stubs like this are customarily ended with an ellipsis instead...

Reply:
I think pass is better because everyone knows what it is.

Reply:
I find `...` is also used in PyTorch, see
https://github.com/pytorch/pytorch/blob/2f895f790a84bffd2aedb1dc184da9837ceb814a/torch/nn/modules/module.py#L704
```python
    @overload
    def to(self: T, device: Optional[Union[int, device]] = ..., dtype: Optional[Union[dtype, str]] = ...,
           non_blocking: bool = ...) -> T:
        ...

    @overload
    def to(self: T, dtype: Union[dtype, str], non_blocking: bool = ...) -> T:
        ...

    @overload
    def to(self: T, tensor: Tensor, non_blocking: bool = ...) -> T:
        ...
```

and
https://github.com/pytorch/pytorch/blob/2f895f790a84bffd2aedb1dc184da9837ceb814a/torch/nn/modules/module.py#L268
```python
    forward: Callable[..., Any] = _forward_unimplemented
```

---

It turns out `...` is an object in Python, similar to `None`. For example,

```python
>>> a = ...
>>> print(a)
Ellipsis
>>> type(a)
<class 'ellipsis'>
>>> b = ...
>>> id(a) == id(b)
True
>>> c = None
>>> print(c)
None
>>> type(None)
<class 'NoneType'>
```

So all of the following definitions have similar effects:

```python
def f(): pass
def f(): ...
def f(): None
def f(): 1
```

though `def f(): pass` is more common.